### PR TITLE
feat(server): Support OIDC custom claims for AuthN. Closes #5953

### LINKED ISSUE #5953: Allow specifying the payload key for groups claim
# Summary

Currently in `v3.0.3`, it is [hardcoding](https://github.com/argoproj/argo-workflows/blob/4e450e250168e6b4d51a126b784e90b11a0162bc/server/auth/types/claims.go#L7) the key `groups` for the `groups` claim.
 It would be very helpful to allow this to be configurable. 

# Use Cases

In our case, our key is `ad_groups` instead of `groups`. This causes argo workflows to not able to read the groups correctly and therefore unable to achieve RBAC SSO via groups claim.

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by liafizan**
Checklist:

* [:heavy_check_mark: ] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

Tips:

* Your PR needs to pass the required checks before it can be approved. If the check is not required (e.g. E2E tests) it does not need to pass
* Sign-off your commits to pass the DCO check: `git commit --signoff`.
* Run `make pre-commit -B` to fix codegen or lint problems. 
* Say how how you tested your changes. If you changed the UI, attach screenshots.

### CODE REVIEW

File: server/info/info_server.go [hunk 1]
```diff
@@ -15,18 +17,64 @@ type infoServer struct {
...
+	if err != nil {
+		return nil, err
+	}
+
+	emailVerified, err := typeassert.Bool(claims["email_verified"])
+	if err != nil {
+		return nil, err
+	}
+
+	serviceAccountName, err := typeassert.String(claims["serviceaccount_name"])
```
Comment (alexec) [inline server/info/info_server.go:64]: wrong

Comment (alexec) [reply server/info/info_server.go:64]: `service_account_name`

**Review by alexec — COMMENTED** [review]
we can simplify this:

```
package main

import (
	"encoding/json"
	"fmt"
)

type Foo struct {
	A int                    `json:"a"`
	B int                    `json:"b"`
	X map[string]interface{} `json:"-"` // Rest of the fields should go here.
}

func main() {
	s := `{"a":1, "b":2, "x":1, "y":1}`
	f := Foo{}
	if err := json.Unmarshal([]byte(s), &f); err != nil {
		panic(err)
	}

	if err := json.Unmarshal([]byte(s), &f.X); err != nil {
		panic(err)
	}

	fmt.Printf("%+v", f)
}

````

In the above example "X" contains all unknown fields, including groups.

File: server/auth/sso/sso.go [hunk 2]
```diff
@@ -232,29 +239,33 @@ func (s *sso) HandleCallback(w http.ResponseWriter, r *http.Request) {
...
+	// (jwt.Builder).Claims expects json convertible input. We convert Claims type
+	// map[string]interface for this
+	jsonClaim, err := jsonutil.Jsonify(c)
 	if err != nil {
 		panic(err)
 	}
+
+	raw, err := jwt.Encrypted(s.encrypter).Claims(jsonClaim).CompactSerialize()
+	if err != nil {
+		panic(err)
```
[the remark this answers is not in the document]
Comment (liafizan) [reply server/auth/sso/sso.go:265]: Thank you but this was already there: https://github.com/argoproj/argo-workflows/blob/master/server/auth/sso/sso.go#L252-L255 and that is why we did not change it thinking there must be some reason why panic was used. As per your suggestion, we will remove this.

[the remark this answers is not in the document]
Comment (alexec) [reply server/auth/sso/sso.go:265]: could we fix it?

**Comment by alexec** [comment]
I think, following my comment, there is a third-way to do this which will result in fewer code changes.

File: USERS.md [hunk 3]
```diff
@@ -135,6 +135,7 @@ Currently, the following organizations are **officially** using Argo Workflows:
 1. [Woolworths Group](https://www.woolworthsgroup.com.au/)
 1. [Workiva](https://www.workiva.com/)
 1. [Zhihu](https://www.zhihu.com/)
+1. [Oracle] (https://www.oracle.com)
```
[the remark this answers is not in the document]
Comment (liafizan) [reply USERS.md]: done

File: server/auth/sso/sso.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (liafizan) [reply server/auth/sso/sso.go:265]: Sure. Going to give it try

**Comment by liafizan** [comment]
> I think, following my comment, there is a third-way to do this which will result in fewer code changes.

Hey @alexec was the last comment in reference to this alternative approach: https://github.com/cocoztho000/argo-workflows/commit/23c9f1628f20d3be06b156e71c157797b328a0be

Just want to be sure that we move in the right direction

**Comment by alexec** [comment]
I think I'd like minimal code changes. This area is hard to test. I think we can do that using the JSON `"-"` option.

**Comment by liafizan** [comment]
> I think I'd like minimal code changes. This area is hard to test. I think we can do that using the JSON `"-"` option.

Thank you for guiding us. We will make changes accordingly and update the PR soon.

**Comment by alexec** [comment]
Sorry for asking for such big change, but I think it'll save as time overall because of fewer bugs.

**Comment by liafizan** [comment]
> Sorry for asking for such big change, but I think it'll save as time overall because of fewer bugs.

Thank you!!. That is a very good point and even we were unsure if we should do that big change and if it's needed. 

Could you please suggest if this approach is OK

```
type Claims struct {
	jwt.Claims
	// We do not remove Groups key but map custom claims to Groups key 
        // this also means, that even if a custom claim is set such as arg_groups, the expr would also use 
        // groups key for evaluation as in:
        // `"admin" in groups` instead of `"admin" in argo_groups`
        // The benefit is that we have very few changes to do with this approach
        // All claims are still available as part of RawClaim and can still be used for edge cases
	Groups             []string               `json:"groups,omitempty"`
	Email              string                 `json:"email,omitempty"`
	EmailVerified      bool                   `json:"email_verified,omitempty"`
	ServiceAccountName string                 `json:"service_account_name,omitempty"`
	RawClaim           map[string]interface{} `json:"-"`
}
```
sso.go 

```
	groups := c.Groups
        // customClaimName is coming from configmap sso.customClaimName field
	if s.customClaimName != "" {
		groups, err = StringSlice(c.RawClaim[s.customClaimName])
		if err != nil {
			w.WriteHeader(401)
			_, _ = w.Write([]byte(fmt.Sprintf("failed to get custom claim: %v", err)))
		}
	}
```
If you think we should get rid of `group` key altogether from Claims, we will take that approach.

**Comment by alexec** [comment]
> Could you please suggest if this approach is OK 

Suck it and see?

**Comment by liafizan** [comment]
> > Could you please suggest if this approach is OK
> 
> Suck it and see?

Assuming this means push, get review & see. We will push changes soon for review. 

Idea was just to get clarity is this is acceptable so that we don't have to make major changes post PR.

**Comment by alexec** [comment]
That would be the goal. Be frugal with the solution.

File: server/auth/sso/sso.go [hunk 4]
```diff
@@ -238,20 +242,36 @@ func (s *sso) HandleCallback(w http.ResponseWriter, r *http.Request) {
...
 	}
+
+	// Default to groups claim but if customClaimName is set
+	// extract groups based on that claim key
+	groups := c.Groups
+	if s.customClaimName != "" {
+		groups, err = c.GetCustomGroup(s.customClaimName)
+		if err != nil {
+			w.WriteHeader(401)
+			_, _ = w.Write([]byte(fmt.Sprintf("failed to get custom claim: %v", err)))
```
Comment (alexec) [inline server/auth/sso/sso.go:253]: return statement missing?

File: server/auth/sso/sso.go [hunk 5]
```diff
@@ -238,20 +242,36 @@ func (s *sso) HandleCallback(w http.ResponseWriter, r *http.Request) {
...
+		RawClaim:           c.RawClaim,
 		Email:              c.Email,
 		EmailVerified:      c.EmailVerified,
 		ServiceAccountName: c.ServiceAccountName,
 	}
+
 	raw, err := jwt.Encrypted(s.encrypter).Claims(argoClaims).CompactSerialize()
 	if err != nil {
-		panic(err)
+		w.WriteHeader(401)
```
Comment (alexec) [inline server/auth/sso/sso.go:272]: 👍

**Review by alexec — APPROVED** [review]

**Pull request merged**